### ISSUE TITLE
Allow rich/custom display of result from R cell (R magic)

### ISSUE THREAD

**Issue by lgautier** [rpy2/rpy2#646]
The result returned by the evalution of an R cell is currenty calling R's own "show" method.
https://github.com/rpy2/rpy2/blob/master/rpy2/ipython/rmagic.py#L750

It would be nice to allow a customization / dispatch according to the type of the output (see issue #641 for example of use-case).

**Comment by richdevboston** [comment]
Yes.  We NEED R magics to work for interoperability.

**Comment by krassowski** [comment]
Another use case would be to allow interactive ggplots enhanced by plotly, e.g. https://plot.ly/ggplot2/animations/.

**Comment by krassowski** [comment]
It seems that the rough idea is that `knitr::knit_print` is almost an equivalent of `IPython.display`, and instead of show, we could default to it (if knitr is installed). Or make it an option.

References:
- [knit_print documentation](https://cran.r-project.org/web/packages/knitr/vignettes/knit_print.html)
- [plotly implementation of knit_print generics](https://github.com/ropensci/plotly/blob/a63b04880a6f1d3104a5e72be728a107bf9c6ef7/R/print.R#L37-L87)
- [gganimate implementation of knit_print generics](https://github.com/thomasp85/gganimate/blob/c5665eca2fe223af1a690f29fa61523292f3a9c0/R/gganim.R#L36-L62)

The knit_print comes with options which are functionally equivalent to the parameters that we pass to R magic cells.

**Comment by lgautier** [comment]
The idea of using `knit_print()` is an enticing one. `knitr` is sufficiently well adopted in the R world to expect continuing and stability. Making this an option would be my first sentiment (since `knitr` is an R package dependency).

Otherwise I have started looking at whether using ipython's `display()` instead of R's `show()` could be a viable solution. This quickly lead me to existing issues with the conversion system, which I need to resolve first.

**Comment by lgautier** [comment]
The current refactoring of the conversion system is in PR #650. There is more work to do though, but the general idea is to allow the dispatch of class wrapping from R to Python by R C-level type + R class name.

However, even with is in place it make sense to consider a print/display dispatcher for R cells that could be specified as an option. I'll open a PR to start this.

**Comment by krassowski** [comment]
This discussion is very relevant: https://github.com/r-lib/pkgdown/issues/1256. It seems that apart from knitr there is a package called repr, which might be of interest.

**Comment by lgautier** [comment]
@krassowski : did you have a chance to check whether the `-d/--display` parameter works for you for the shorter term ? This would be included in release 3.3.0.

**Comment by krassowski** [comment]
Looking at it right now, on a fresh install from master. gganimate still gives me a hard time - if I print the returned object in a callback function specified with `-d` it raises an error as reported previously (something in repr I guess, will follow up later), but I can otherwise operate on the object and the `-d` option works well.

Here is a recording from my first attempt at utilizing it:

![gganimate_optimized](https://user-images.githubusercontent.com/5832902/77861164-453c2e00-720b-11ea-960d-7977b18058fe.gif)

```python
%load_ext rpy2.ipython
from IPython.display import Image, display
from rpy2.robjects.packages import importr
gganimate = importr('gganimate')

def embed_gif(animation):
    path = gganimate.animate(
        animation, renderer=gganimate.gifski_renderer(),
        width=350, height=300, res=90
    )
    with open(list(path)[0], 'rb') as f:
        display(Image(data=f.read(), format='gif'))

```

```R
%%R --display embed_gif
library(ggplot2)
library(gganimate)

ggplot(mtcars, aes(factor(cyl), mpg)) + 
  geom_boxplot() + 
  transition_states(
    gear,
    transition_length = 2,
    state_length = 1
  )
```

**Comment by krassowski** [comment]
It seems ok as a temporary solution. Would it make sense to pass keyword arguments from the cell to the display function (such as width, height, dpi?).

Thank you for addressing this and other issues so quickly!

**Comment by lgautier** [comment]
> It seems ok as a temporary solution. Would it make sense to pass keyword arguments from the cell to the display function (such as width, height, dpi?).

Yes. Graphical parameters should be accessible to that function.

**Comment by lgautier** [comment]
Otherwise it looks it might worthwhile to modify the callback for R writing to the console so that you have less text verbosity because of the progress bar.

**Comment by lgautier** [comment]
PR#671 is proposing to just pass all arguments given to the magic / R cell. Would that work ?

**Comment by richdevboston** [comment]
The R function output into the JuPyter output cell should NOT be the default setting.  It makes Jupyter documents excessively long to always include console output.  However, an option to show progress bars in console output for debugging purposes will be helpful but this should be OFF as "the sensible default".  Thank you.

**Comment by krassowski** [comment]
Just following up on this, the code below makes use of the args to customise the display (and it works well for me):

```python
%load_ext rpy2.ipython
from IPython.display import Image, display
from rpy2.robjects.packages import importr
gganimate = importr('gganimate')


def embed_gif(animation, args):
    path = gganimate.animate(
        animation, renderer=gganimate.gifski_renderer(),
        width=args.width, height=args.height, res=args.res
    )
    with open(list(path)[0], 'rb') as f:
        display(Image(data=f.read(), format='gif'))
```

```R
%%R --display embed_gif  -w 350 -h 300 -r 90
library(ggplot2)
library(gganimate)

ggplot(mtcars, aes(factor(cyl), mpg)) + 
  geom_boxplot() + 
  transition_states(
    gear,
    transition_length = 2,
    state_length = 1
  )
```

GIF recording (sorry about the blurred text, these are Google Chrome + HDPI + Ubuntu issues)

![rpy2_gganimate_all_good](https://user-images.githubusercontent.com/5832902/82549947-aa4e3900-9b55-11ea-9e52-95f23cdb525c.gif)

**Comment by joelostblom** [comment]
First of all, thanks for making `rpy2`! I have just started using it and it has already been very helpful!

I wonder if the `-d` parameter could be used to display `plotly` figures inside JupyterLab? For example, the code below outputs an interactive plot when I run the notebook with an R kernel. 

```R
library(plotly)

fig <- plot_ly(midwest, x = ~percollege, color = ~state, type = "box")
fig
```

However, when I run with an Python kernel and the `%%R` cell magic, there is no plot showed and an error tell me to set the browser string. When I do that (`options('browser' = 'chromium')`), the figures render correctly in a separate browser tab instead of showing up inline in the notebook (the 'browser' option is not set when using and R kernel, so I don't think it is needed).

I tried using `display`, `display_html` and `HTML` from the `IPython.display` package as well as `show` from `plotly.io` and `FigureWidget` from `plotly.graph_objects`, but neither of these work when I pass it to `-d` (they give a few different errors which I can list here if preferred but the output is rather verbose).

The [plotly docs mention that htmlwidgets is used to display plotly by default in R-plotly](https://plotly.com/r/getting-started/), so I am not sure if the Jupyter notebook hooks into this somehow when using the R kernel directly.

Do you have any suggestions on how to make `-d` work to display interactive plotly plots?

**Comment by krassowski** [comment]
Yes, I think it should be possible. I just managed to achieve something very similar with [ggiraph](https://github.com/davidgohel/ggiraph) for interactive plots.
As both ggiraph and plotly use htmlwidgets you should be able to adapt my ggiraph code snippet (below). Thinking about it, it could even work as is!

```python
%load_ext rpy2.ipython
from pathlib import Path
from IPython.display import HTML, display
from rpy2.robjects.packages import importr

base = importr('base')
htmlwidgets  = importr('htmlwidgets')

def girafe(interactive_plot, args):
    knitr_output = htmlwidgets.knit_print_htmlwidget(interactive_plot, standalone=False)
    html = list(knitr_output)[0].replace('<!--html_preserve-->', '').replace('<!--/html_preserve-->', '')
    deps = base.attr(knitr_output, 'knit_meta')
    javascript = ''
    css = ''
    for dep in deps:
        path = Path(list(dep.rx2('src').rx2('file'))[0])
        stylesheet = dep.rx2('stylesheet')
        if stylesheet:
            stylesheet = list(stylesheet)[0]
            content = (path/stylesheet).read_text()
            css += f'{content}\n'
        script = dep.rx2('script')
        if script:
            script = list(script)[0]
            content = (path/script).read_text()
            javascript += f'{content}\n\n'
    style = ''
    for arg in ['width', 'height']:
        value = getattr(args, arg)
        if value is not None:
            style += f'{arg}: {value}px;'
    invoker = (
        "let girafe = this.closest('.girafe');"
        "let scripts = girafe.querySelectorAll('script');"
        "scripts.forEach(function(script){ if (script.type == 'text/javascript') eval(script.innerHTML) });"
        "window.HTMLWidgets.staticRender();"
        "this.parentNode.removeChild(this)"
    )
    display(
        HTML(
            data=(
                f'<div class="girafe">'
                f'<script type="text/javascript">{javascript}</script>'
                f'<div style="{style}">{html}</div>'
                f'<img src onerror="{invoker}">'
                f'</div>'
            )
        )
    )
```

and:

```R
%%R --display girafe -w 600

library(ggplot2)
library(ggiraph)
g <- ggplot(mpg, aes(x = displ, y = cty, color = hwy))

my_gg <- g + geom_point_interactive(aes(tooltip = model), size = 2)
girafe(code = print(my_gg))
```

**Comment by krassowski** [comment]
I should have mentioned, it was only tested with JupyterLab. Demo:

![ggiraph](https://user-images.githubusercontent.com/5832902/95666720-59ff6400-0b54-11eb-9fce-2caee589c2e5.gif)

**Comment by joelostblom** [comment]
Cool! Thanks for sharing @krassowski ! It works for me when I run your `ggiraph` code, but when I run the plotly code I posted above I get
```
FileNotFoundError: [Errno 2] No such file or directory: 'htmlwidgets/lib/typedarray/typedarray.min.js'
```
Would you have any idea where to start looking to solve this?

Full traceback:

```
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-3-9d676bc98d00> in <module>
----> 1 get_ipython().run_cell_magic('R', '-d girafe', '\nlibrary(plotly)\n\nfig <- plot_ly(midwest, x = ~percollege, color = ~state, type = "box")\nfig\n')

~/miniconda3/envs/jl/lib/python3.8/site-packages/IPython/core/interactiveshell.py in run_cell_magic(self, magic_name, line, cell)
   2379             with self.builtin_trap:
   2380                 args = (magic_arg_s, cell)
-> 2381                 result = fn(*args, **kwargs)
   2382             return result
   2383 

<decorator-gen-121> in R(self, line, cell, local_ns)

~/miniconda3/envs/jl/lib/python3.8/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188 
    189         if callable(arg):

~/miniconda3/envs/jl/lib/python3.8/site-packages/rpy2/ipython/rmagic.py in R(self, line, cell, local_ns)
    773                                                'consolewrite_print',
    774                                                self.write_console_regular))
--> 775                         cell_display(result, args)
    776                         text_output += self.flush()
    777 

<ipython-input-1-be04556a1afa> in girafe(interactive_plot, args)
     23         if script:
     24             script = list(script)[0]
---> 25             content = (path/script).read_text()
     26             javascript += f'{content}\n\n'
     27     style = ''

~/miniconda3/envs/jl/lib/python3.8/pathlib.py in read_text(self, encoding, errors)
   1233         Open the file in text mode, read it, and close the file.
   1234         """
-> 1235         with self.open(mode='r', encoding=encoding, errors=errors) as f:
   1236             return f.read()
   1237 

~/miniconda3/envs/jl/lib/python3.8/pathlib.py in open(self, mode, buffering, encoding, errors, newline)
   1219         if self._closed:
   1220             self._raise_closed()
-> 1221         return io.open(self, mode, buffering, encoding, errors, newline,
   1222                        opener=self._opener)
   1223 

~/miniconda3/envs/jl/lib/python3.8/pathlib.py in _opener(self, name, flags, mode)
   1075     def _opener(self, name, flags, mode=0o666):
   1076         # A stub for the opener argument to built-in open()
-> 1077         return self._accessor.open(self, flags, mode)
   1078 
   1079     def _raw_open(self, flags, mode=0o777):

FileNotFoundError: [Errno 2] No such file or directory: 'htmlwidgets/lib/typedarray/typedarray.min.js'
```

**Comment by krassowski** [comment]
I edited my script above to handle specification of files better and it's working for me. Also renamed the function to `htmlwidget` :)

![plotly](https://user-images.githubusercontent.com/5832902/95677773-b51a7080-0bbf-11eb-8ea8-f4440ff73e8b.gif)

Let me know how it works for you.

@lgautier would it make sense to distribute `embed_gif` and `htmlwidget` as an extension to rpy2? Something like `rpy2-ipython-displays` or `rpy2-ipython-renderers`?

**Comment by richdevboston** [comment]
Please be careful about taking the name HTMLwidget for you function,
because I think there is already a library by that name.  Perhaps being
more specific about the name of your function can avoid the collision.

Ryszard

**Comment by joelostblom** [comment]
This works!! :rocket:  Thank you so much for taking the time to generalize it! I would love to see these display functions become built into `rpy2` to make them more accessible.

@richdevboston I believe the library has an `s` at the end and if this display function works with figure that are using that library in general, I would say the name is quite suitable (my non-expert opinion).

**Comment by lgautier** [comment]
> @lgautier would it make sense to distribute embed_gif and htmlwidget as an extension to rpy2? Something like rpy2-ipython-displays or rpy2-ipython-renderers?

@krassowski . Works for me. I have been working on creating an umbrella `rpy2` on GitHub, albeit slowly: https://github.com/rpy2 (adding references to them from the main is also easily, although this still needs a bit of work on the aesthetics - https://rpy2.github.io/extensions.html).

I am happy to create a repos there for you

**Comment by lgautier** [comment]
> This works!! rocket Thank you so much for taking the time to generalize it! I would love to see these display functions become built into rpy2 to make them more accessible

@joelostblom : over time rpy2's user base grew a fair bit (~1M downloads / year from pypi) while remaining a fairly small operation. The current thinking is to have extensions under the management of their respective authors. With a bit of help from automated continuous integration, and extension setups to use as templates for new efforts, this will let extensions have very short release cycles as they establish themselves.

**Comment by krassowski** [comment]
Happy it worked! The function can certainly be improved and I will be happy to work on it in the new repo. @lgautier, yes please.

**Comment by lgautier** [comment]
@krassowski : what name would you like to have for the repos?

**Comment by krassowski** [comment]
Maybe `rpy2-ipython-display`? Could be either plural or singular (plural kind of makes sense at first glance, though plain IPython's "equivalent" is `IPython.display`, so this could be more intuitive).

**Comment by lgautier** [comment]
@krassowski : I created this and made you admin of  it - https://github.com/rpy2/rpy2-ipython-display

I also thought later having that repos under your account is also a possibility. I am not saying that you should, just that there are many options and I am equally comfortable with them. ;-)

**Comment by krassowski** [comment]
Got it, thank you! I prefer it to be in this organization, seems like a better fit there :)

I will start slowly (busy month), but hope to get something publishable before November.

**Comment by joelostblom** [comment]
I'm trying to use the `-d/--display` magic to show tibbles/dataframes from R as rich HTML tables in the notebook (how pandas dataframes display by default). I naively tried to use `html_rdataframe` without success:

![image](https://github.com/user-attachments/assets/7d00d2d0-b324-491e-8f37-be3def651c54)

Is there currently a way to display R dataframes as HTML tables (either with or without the `-d` magic option?

**Comment by lgautier** [comment]
A change must have broken with the conversion, and no test caught it. I'd have to look more closely at what a proper fix should be.

In the meantime the following should bring the built-in HTML rendering back:

```python
import rpy2.robjects.conversion
import IPython.display

def html_display(x, *args):
    with rpy2.robjects.conversion.get_conversion().context() as ctx:
        res = ctx.rpy2py(x)
    IPython.display.display(res)
```

```python
%%R -d html_displays

mtcars
```

Alternatively, you could output the `data.frame` of interest. It will be converted to a pandas `DataFrame`:

```python
%%R -o mtcars -d html_silent

mtcars
```

```python
# Show a pandas `DataFrame`.
mtcars
```

**Comment by joelostblom** [comment]
Thanks for the workaround! It seems like using `html_display` leaves out the rownames whereas they are shown if exporting the dataframe to a Python cell and showing it via pandas, or viewing the dataframe in a notebook with an R kernel.
